### PR TITLE
feat(rnsdk) Create initial directory for React Native SDK

### PR DESCRIPTION
**Pull request by filiprejmus**
Create the directory which will be the React Native SDK including its podspec and package.json. Also add a new script to move the required files to the sdk directory and add it to package.json

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -198,7 +199,8 @@
     "lint-fix": "eslint --ext .js,.ts,.tsx --max-warnings 0 --fix .",
     "postinstall": "patch-package --error-on-fail && jetify",
     "validate": "npm ls",
-    "start": "make dev"
+    "start": "make dev",
+    "prepare": "sh ./prepare_sdk.sh"
```
Comment (saghul) [inline package.json:203]: Let's put this in the SDK project's package.json file.

File: package.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
   "dependencies": {
     "@amplitude/react-native": "2.7.0",
     "@atlaskit/button": "15.1.4",
+    "@atlaskit/checkbox": "12.0.0",
```
Comment (saghul) [inline package.json:20]: This looks like a merge error, it's now gone.

File: prepare_sdk.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (saghul) [inline prepare_sdk.sh:1]: Please move the script to the rnsdk folder and run it within.

File: package.json (hunk 1 above)
Comment (filiprejmus) [reply package.json]: Done

File: package.json (hunk 2 above)
Comment (filiprejmus) [reply package.json]: Fixed

File: prepare_sdk.sh (hunk 3 above)
Comment (filiprejmus) [reply prepare_sdk.sh]: Done

**Comment by filiprejmus** [comment]
new changes ready for review :)

**Comment by Aleksandern** [comment]
Hi @filiprejmus 
Could you describe how to use that?

**Comment by filiprejmus** [comment]
It's a couple of steps. I'll try to minimize it.
1.) Clone https://github.com/filiprejmus/JitsiMeetSDKTester
2.) On this branch here, go into the react-native-sdk folder and `npm run prepare && npm pack`
3.) npm install [path where you ran pack]/[the name of the .tgz file]
4.) cd ios && pod install && cd ..
5.) Open the XCode Workspace of SDKTester
6.) Under Pod, Target jitsi-meet-react-native link the binary with the WebRTC.xcframework, if it's not available run `SDKTester/node_modules/react-native-webrtc/tools/downloadBitcode.sh`
7.) Under SDK Tester Info add Privacy - Microphone Usage Description and Privacy - Camera Usage Description

That should do to make it run, assuming the Callkit situation is fixed

**Comment by Aleksandern** [comment]
@filiprejmus 
Thank you for your explanation. Will it work for Android?

**Comment by filiprejmus** [comment]
> @filiprejmus 
> Thank you for your explanation. Will it work for Android?

It will, it's wip

**Comment by Aleksandern** [comment]
@filiprejmus 
It seems to be work okay on iOS but I don't have the "share screen" button.  Maybe you know how to enable the share screen feature?

**Comment by Aleksandern** [comment]
> @filiprejmus
> It seems to be work okay on iOS but I don't have the "share screen" button. Maybe you know how to enable the share screen feature?

@filiprejmus 
`flags` props must be Object but you described it as Array.

But the share screen feature doesn't work anyway.

**Comment by saghul** [comment]
Screen sharing requires a broadcast extension to be implemented by the app.

That is out of scope of this project for now.

**Comment by Aleksandern** [comment]
@saghul 
> Screen sharing requires a broadcast extension to be implemented by the app.

Will the screen sharing work if I implement the broadcast extension to my app using this doc  [screen-sharing-integration](https://jitsi.github.io/handbook/docs/dev-guide/dev-guide-ios-sdk#screen-sharing-integration) or needs to be done some additional steps for `react-native-sdk-directory` ?

**Comment by saghul** [comment]
> @saghul
> 
> > Screen sharing requires a broadcast extension to be implemented by the app.
> 
> Will the screen sharing work if I implement the broadcast extension to my app using this doc [screen-sharing-integration](https://jitsi.github.io/handbook/docs/dev-guide/dev-guide-ios-sdk#screen-sharing-integration) or needs to be done some additional steps for `react-native-sdk-directory` ?

Probably not in the first iteration since there are some extra bits required that we'll look into once we get the  SDK off the ground.

File: react/features/app/components/App.native.js [hunk 4]
```diff
@@ -140,6 +140,7 @@ export class App extends AbstractApp {
         }
 
         dispatch(updateSettings(this.props.userInfo || {}));
+        dispatch(updateSettings(this.props.settings || {}));
```
Comment (saghul) [inline react/features/app/components/App.native.js:143]: Why is this necessary?

File: react/features/mobile/react-native-sdk/functions.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Dispatches a READY_TO_CLOSE event from an arbitrary component.
+ * Created to enable triggering READY_TO_CLOSE from a SDK component.
+ *
+ * @param {Function} dispatch - The dispatch function.
+ * @returns {void}
+ */
+export function closeJitsiMeeting(dispatch) {
+    _sendReadyToClose(dispatch);
```
Comment (saghul) [inline react/features/mobile/react-native-sdk/functions.js:29]: You shouldn't use this directly, it's dispatched as a side-effect. Instead, in order to leave a meeting do `dispatch(appNavigate(undefined)` and that will tear down the meeting and dispatch ready to close.

File: react/features/mobile/react-native-sdk/middleware.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+const externalAPIEnabled = isExternalAPIAvailable();
+
+/**
+ * Check if native modules are being used or not. If not then the init of middleware doesn't happen.
+ */
+!externalAPIEnabled && MiddlewareRegistry.register(store => next => action => {
+    const result = next(action);
+    const { type } = action;
+    const leave = getAppProp(store, 'onReadyToClose');
```
Comment (saghul) [inline react/features/mobile/react-native-sdk/middleware.js:20]: why not call it onReadyToClose for consistency with the other events?

File: react/features/mobile/react-native-sdk/middleware.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    const onConferenceJoined = getAppProp(store, 'onConferenceJoined');
+    const onConferenceWillJoin = getAppProp(store, 'onConferenceWillJoin');
+    const onConferenceLeft = getAppProp(store, 'onConferencLeft');
```
Comment (saghul) [inline react/features/mobile/react-native-sdk/middleware.js:23]: Can you please goup all of those props into a single object like `rnSdkHandlers` or something?

File: react-native-sdk/index.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
+import 'react-native-gesture-handler';
```
Comment (saghul) [inline react-native-sdk/index.js:1]: I see there is some duplicity here with index.native.js, how about moving the root component in that file to some Root.js and then importing it here and there?

File: react-native-sdk/index.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            startWithVideoMuted?: boolean;
+            startWithReactionsMuted?: boolean;
+        }
+    }
+};
+
+/**
+ * Main React Native SDK component that displays a Jitsi Meet conference and gets all required params as props
+ */
+const JitsiMeetView = forwardRef((props: AppProps, ref) => {
```
Comment (saghul) [inline react-native-sdk/index.js:42]: Instead of calling it this way, let's try to align it a bit with the react SDK and call it JitsiMeeting.

File: react-native-sdk/index.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        roomName: string;
+        onReadyToClose?: Function;
+        onConferenceJoined?: Function;
+        onConferenceWillJoin?: Function;
+        onConferenceLeft?: Function;
+        settings?: {
+            startWithAudioMuted?: boolean;
+            startAudioOnly?: boolean;
+            startWithVideoMuted?: boolean;
+            startWithReactionsMuted?: boolean;
```
Comment (saghul) [inline react-native-sdk/index.js:34]: We don't have a way to toggle this on the native side yet, so please skip it.

File: react-native-sdk/index.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                'onConferenceLeft': props.meetingOptions.onConferenceLeft,
+                'flags': props.flags,
+                'settings': props.meetingOptions.settings
+            });
+        }, []
+    );
+
+
+    return (
+        <View style={{ width: props.width,
```
Comment (saghul) [inline react-native-sdk/index.js:79]: Have a `style` prop instead, so give some extra flexibility.

**Review by saghul — COMMENTED** [review]
Left some comments, please take a look!

File: react/features/app/components/App.native.js (hunk 4 above)
Comment (saghul) [reply react/features/app/components/App.native.js:143]: Those are passed as config overrides, and the config middleware takes care of the rest.

Check the asProps methods on the native SSK to see how those are passed.

File: react-native-sdk/index.js (hunk 8 above)
Comment (filiprejmus) [reply react-native-sdk/index.js:1]: We're using a ref in the sdk but not in the index.native. In my opinion the reduction of duplications is worth less than the implementation of another ref forwarding component that also only sometimes needs the ref.

Comment (saghul) [reply react-native-sdk/index.js:1]: OK, let's leave it for now and maybe we come up with other ideas.

**Comment by kahx** [comment]
> > @filiprejmus
> > Thank you for your explanation. Will it work for Android?
> 
> It will, it's wip

Is there any date for Android or complete release?

**Comment by filiprejmus** [comment]
> > > @filiprejmus
> > > Thank you for your explanation. Will it work for Android?
> > 
> > 
> > It will, it's wip
> 
> Is there any date for Android or complete release?

The GSOC project is scheduled to end mid October. If everything goes smoothly I expect the npm release to be in that timespan. Final decision lies with @saghul though

**Comment by ratz6** [comment]
are native modules used here or the entire package is done in react native ? 

I'm using lib-jitsi-meet in react and using a webView in react native . gives me a lot of freedom over layouts and features but eagerly waiting for something like this in react native itself to show up.

Let me know if there's any help req

**Comment by filiprejmus** [comment]
> are native modules used here or the entire package is done in react native ?
> 
> I'm using lib-jitsi-meet in react and using a webView in react native . gives me a lot of freedom over layouts and features but eagerly waiting for something like this in react native itself to show up.
> 
> Let me know if there's any help req

There's native modules, they are listed in prepare_sdk.js

**Comment by filiprejmus** [comment]
@Aleksandern @ratz6 
As you offered help in case there is need:
I am having trouble linking a framework dependency. I described the problem in detail [here](https://stackoverflow.com/questions/73825977/link-binary-with-library-from-podspec).

**Comment by jean-ben-guigui** [comment]
Hi, I am very much interested in this module, is everything going well? Do you think you'll be able to deliver mid October as planned? Thanks for the great work 😀

**Comment by filiprejmus** [comment]
> 

I believe end of October is a realistic timeline for a working SDK!

**Comment by mohit2546** [comment]
When can we expect a working sdk waiting for this from a long time and thanks for the great work

**Comment by filiprejmus** [comment]
> When can we expect a working sdk waiting for this from a long time and thanks for the great work

It's really just a couple more refactorings on android side and some final clean ups, so really soon!

**Comment by mohit2546** [comment]
@filiprejmus thank you for your response appreciate it  🙌

**Comment by GleidsonDaniel** [comment]
@filiprejmus What is missing for this new sdk to be released?
Here at Rocket.Chat Jitsi is one of the sdks used in the project and this lack of a specific sdk for the RN is one of the gaps for the project (we even thought about dropping the use of native sdk and trying to have calls only via webview, because the support is pretty bad), but if this new sdk is released it's a great point to consider.
Today millions of users use the rocket.chat app around the world and as this new sdk is something very important.
My question is not a charge but a statement of help, if something is missing what can we do to help get this sdk out?

**Comment by saghul** [comment]
> @filiprejmus What is missing for this new sdk to be released?

It's in our hands now. We've been quite busy lately and haven't been able to wrap it up.



> My question is not a charge but a statement of help, if something is missing what can we do to help get this sdk out?

Thank you! It would be quite useful if you could try to give it a try in its current shape.

**Comment by filiprejmus** [comment]
@GleidsonDaniel I created a demo repository where you can find a example [here](https://github.com/filiprejmus/JitsiSDKTester)
Note that you'll need to checkout this branch and build a npm package locally before you can use it

**Comment by GleidsonDaniel** [comment]
> 

@filiprejmus you talk about building the sdk from your branch here: https://github.com/filiprejmus/jitsi-meet/tree/create-react-native-sdk-directory

If so, I'll give it a test here, hopefully it works.
It will be a great advance.

**Comment by GleidsonDaniel** [comment]
@filiprejmus well I published the package, basically I cloned your repo and published it on npm without any changes, I also cloned your other repository, which you linked in the previous message.
But it returns some errors... Including one of them is one that is reported here: https://github.com/jitsi/jitsi-meet/issues/12464

One related to environment variables (Which are simple to resolve)
But the other is something related to the SDK, that's the reason for posting here and not trying to solve it hahaha

log:
```
error Failed to install the app. Make sure you have the Android development environment set up: https://reactnative.dev/docs/environment-setup.
Error: Command failed: ./gradlew app:installDebug -PreactNativeDevServerPort=8081
/Users/daniel/Documents/dev/JitsiSDKTester/node_modules/jitsi-meet-react-native-rocket/android/src/main/java/org/jitsi/meet/sdk/AppInfoModule.java:78: error: cannot find symbol
        constants.put("LIBRE_BUILD", BuildConfig.LIBRE_BUILD);
                                                ^
  symbol:   variable LIBRE_BUILD
  location: class BuildConfig
/Users/daniel/Documents/dev/JitsiSDKTester/node_modules/jitsi-meet-react-native-rocket/android/src/main/java/org/jitsi/meet/sdk/AppInfoModule.java:79: error: cannot find symbol
        constants.put("GOOGLE_SERVICES_ENABLED", BuildConfig.GOOGLE_SERVICES_ENABLED);
                                                            ^
  symbol:   variable GOOGLE_SERVICES_ENABLED
  location: class BuildConfig
/Users/daniel/Documents/dev/JitsiSDKTester/node_modules/jitsi-meet-react-native-rocket/android/src/main/java/org/jitsi/meet/sdk/ConnectionService.java:360: error: cannot find symbol
            ReactInstanceManagerHolder.emitEvent(
            ^
  symbol:   variable ReactInstanceManagerHolder
  location: class ConnectionService.ConnectionImpl
/Users/daniel/Documents/dev/JitsiSDKTester/node_modules/jitsi-meet-react-native-rocket/android/src/main/java/org/jitsi/meet/sdk/ConnectionService.java:380: error: cannot find symbol
            ReactInstanceManagerHolder.emitEvent(
            ^
  symbol:   variable ReactInstanceManagerHolder
  location: class ConnectionService.ConnectionImpl
/Users/daniel/Documents/dev/JitsiSDKTester/node_modules/jitsi-meet-react-native-rocket/android/src/main/java/org/jitsi/meet/sdk/ConnectionService.java:409: error: cannot find symbol
            RNConnectionService module = ReactInstanceManagerHolder.getNativeModule(RNConnectionService.class);
                                         ^
  symbol:   variable ReactInstanceManagerHolder
  location: class ConnectionService.ConnectionImpl
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
5 errors
```

Including @saghul, your question is answered in this comment.
Clone this repo, publish the package to NPM, clone the test repo and try to run the project.

**Comment by KolissnikBogdan** [comment]
@filiprejmus Hi, maybe you can tell at least the approximate release date of this sdk?)

**Comment by UltWolf** [comment]
Hello, When can we expect a working SDK? Waiting for this for a long time. Thanks for the great work.

**Comment by saghul** [comment]
Sorry, the year end is being busier than expected. We'll get to wrapping this up as soon as we can.

**Comment by ankar84** [comment]
> Sorry, the year end is being busier than expected. We'll get to wrapping this up as soon as we can.

Wait that a lot, @saghul 
And thanks for your great work, Saul!

**Comment by vinayakathawle** [comment]
> Sorry, the year end is being busier than expected. We'll get to wrapping this up as soon as we can.

@saghul - have you release skd for react native ? I need help to create RN app for Android with Jitsi latest sdk.
Please guide me.

I was looking it to this - https://github.com/filiprejmus/JitsiSDKTester but it has local dir for jitsi react native sdk.

Thanks,

**Comment by keremoge** [comment]
This work will be great for **react-native** developers

**Comment by shashwat-concent** [comment]
Any update for alpha release with bare minimum call connect disconnect video? I am sure people can contribute to the features part?

**Comment by saghul** [comment]
Folks, please stop asking or I'll have to lock this issue. It will be ready when it's ready.

File: react-native-sdk/android/build.gradle [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      url defaultDir.toString()
+      name androidSourcesName
+    }
+
+    logger.info(":${project.name}:reactNativeAndroidRoot ${defaultDir.canonicalPath}")
+    found = true
+  } else {
+    def parentDir = rootProject.projectDir
+
+    1.upto(5, {
```
Comment (tmoldovan8x8) [inline react-native-sdk/android/build.gradle:82]: Not sure I understand why this is done 5 times. Could you maybe leave a comment?

File: react-native-sdk/android/gradle.properties [hunk 13]
```diff
@@ -0,0 +1,5 @@
+JitsiMeetReactNative_kotlinVersion=1.7.0
+JitsiMeetReactNative_minSdkVersion=21
+JitsiMeetReactNative_targetSdkVersion=31
+JitsiMeetReactNative_compileSdkVersion=31
+JitsiMeetReactNative_ndkversion=21.4.7075529
```
Comment (tmoldovan8x8) [inline react-native-sdk/android/gradle.properties:5]: Is this ok for M1 users?

File: react/features/mobile/external-api/functions.js [hunk 14]
```diff
@@ -27,3 +27,21 @@ export function sendEvent(store: Object, name: string, data: Object) {
...
     dispatch(readyToClose());
 }, 2500, { leading: true });
+
+/**
+ * Returns a participant info object based on the passed participant object from redux.
+ *
+ * @param {Participant} participant - The participant object from the redux store.
+ * @returns {Object} - The participant info object.
+ */
+export function _participantToParticipantInfo(participant) {
```
Comment (tmoldovan8x8) [inline react/features/mobile/external-api/functions.js:37]: This is not private anymore, you can loose the _

File: react/features/mobile/react-native-sdk/middleware.js [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+ * Check if native modules are being used or not. If not then the init of middleware doesn't happen.
+ */
+!externalAPIEnabled && MiddlewareRegistry.register(store => next => action => {
+    const result = next(action);
+    const { type } = action;
+    const rnSdkHandlers = getAppProp(store, 'rnSdkHandlers');
+
+    switch (type) {
+    case READY_TO_CLOSE:
+        rnSdkHandlers.onReadyToClose && rnSdkHandlers.onReadyToClose();
```
Comment (tmoldovan8x8) [inline react/features/mobile/react-native-sdk/middleware.js:26]: Maybe use ?.()

File: react/features/mobile/react-native-sdk/functions.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Dispatches a READY_TO_CLOSE event from an arbitrary component.
+ * Created to enable triggering READY_TO_CLOSE from a SDK component.
+ *
+ * @param {Function} dispatch - The dispatch function.
+ * @returns {void}
+ */
+export function closeJitsiMeeting(dispatch) {
+    dispatch(appNavigate(undefined));
```
Comment (tmoldovan8x8) [inline react/features/mobile/react-native-sdk/functions.js:28]: Is this required for READY_TO_CLOSE event only or the actual navigation to undefined is required?

**Comment by saghul** [comment]
@tmoldovan8x8 are your last comments for us? I don't think it's reasonable expect @filiprejmus to pick this back up :-) 

Let's amend and push it over the finish line.

**Comment by ratz6** [comment]
@saghul @filiprejmus where are we with this ? is it stopped ? no more RN SDK ? what's going on ??

**Comment by chr314** [comment]
@ratz6 #13060
